File: Objective-C/CT/Colors Trace/Handlers/ColorCanvasHandler.m
//
//  ColorCanvasHandler.m
//  CT
//
//  Created by Development on 01.07.2022.
//

#import "ColorCanvasHandler.h"
#import "Stack.h"

@interface ColorCanvasHandler ()
@property (strong, nonatomic) UIView *canvas;
@property (nonatomic, copy) void (^wasChanged)(BOOL isUndoEnabled, BOOL isRedoEnabled);

@property (strong, nonatomic) UIColor *selectedColor;
@property (strong, nonatomic) NSMutableDictionary *hashColorDictionary;
@property (strong, nonatomic) Stack *undoStack;
@property (strong, nonatomic) Stack *redoStack;
@end

@implementation ColorCanvasHandler

// Override Setters:
- (void)setCanvas:(UIView *)canvas {
    // A Boolean indicating whether sublayers are clipped to the layer’s bounds
    [canvas.layer setMasksToBounds:true];
    // Set the canvas
    _canvas = canvas;
    [self addTapGestureRecognizerToCanvas];
}

- (id)initWithCanvas:(UIView *) canvas
withWasChangedHandler:(void (^)(BOOL isUndoEnabled, BOOL isRedoEnabled)) wasChangedHandler {
    self = [super init];
    if (self) {
        self.canvas = canvas;
        self.wasChanged = wasChangedHandler;
        
        // Init dictionary for hash colors
        self.hashColorDictionary = [[NSMutableDictionary alloc] init];
        // Init undo stack
        self.undoStack = [[Stack alloc] init];
        // Init undo stack
        self.redoStack = [[Stack alloc] init];
    }
    return self;
}

- (void)addTapGestureRecognizerToCanvas {
    UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget: self action:@selector(addView:)];
    [self.canvas addGestureRecognizer:tapGesture];
}

- (void)addView:(UITapGestureRecognizer *) gesture {
    if (self.selectedColor) {
        // Get location on canvas
        CGPoint location = [gesture locationInView:self.canvas];
        UIColor *color = self.selectedColor;
        // Add view at location on canvas with specific color
        [self addViewAt:location withColor:color];
    }
}

- (void)addViewAt:(CGPoint) location withColor:(UIColor *) color {
    CGFloat size = 50.0;
    UILabel *view = [[UILabel alloc] initWithFrame:CGRectMake(location.x - size * 0.5, location.y - size * 0.5, size, size)];
    view.backgroundColor = color;
    // Add rounded corners
    view.layer.cornerRadius = size * 0.5;
    view.clipsToBounds = true;
    // Set font
    [view setFont:[UIFont systemFontOfSize:20]];
    // Set text attributes
    [view setTextColor:[UIColor whiteColor]];
    [view setTextAlignment:NSTextAlignmentCenter];
    // Update hash for added color
    int current = [self updateHashWithColor:color withAscending:true];
    [view setText:[NSString stringWithFormat:@"%i",current]];
    // Add to canvas
    [self.canvas addSubview:view];
    // Add view to "undo" stack
    [self.undoStack push:view];
    // Completion handler
    BOOL isUndoEnabled = [self.undoStack.items count] == 0 ? false : true;
    BOOL isRedoEnabled = [self.redoStack.items count] == 0 ? false : true;
    self.wasChanged(isUndoEnabled, isRedoEnabled);
}

- (int)updateHashWithColor:(UIColor *) color withAscending:(BOOL) isAscending {
    int current = [[self.hashColorDictionary objectForKey:color] intValue];
    current = isAscending ? current + 1 : MAX(0, current - 1);
    [self.hashColorDictionary setObject:[NSNumber numberWithInt:current] forKey:color];
    return current;
}

- (void)colorWasChangedWith:(UIColor *)color {
    self.selectedColor = color;
}

- (void)undo {
    UIView *view = [self.undoStack pop];
    // Update hash for removed color
    [self updateHashWithColor: view.backgroundColor withAscending: false];
    // Remove view from canvas
    [view removeFromSuperview];
    // Add view to "redo" stack
    [self.redoStack push:view];
    // Completion handler
    BOOL isUndoEnabled = [self.undoStack.items count] == 0 ? false : true;
    BOOL isRedoEnabled = [self.redoStack.items count] == 0 ? false : true;
    self.wasChanged(isUndoEnabled, isRedoEnabled);
}

- (void)redo {
    UIView *view = [self.redoStack pop];
    // Get location from restored view
    CGPoint location = CGPointMake(view.frame.origin.x + view.frame.size.width * 0.5, view.frame.origin.y + view.frame.size.height * 0.5);
    // Get color from restored view
    UIColor *color = view.backgroundColor;
    // Add restored view at location on canvas with specific color
    [self addViewAt:location withColor:color];
}

@end

File: SwiftUIKit/CT/Colors Trace/Handlers/ColorCanvasHandler.swift
//
//  ColorCanvasHandler.swift
//  CT
//
//  Created by Development on 01.07.2022.
//

import Foundation
import UIKit

/// An object that hold canvas for displaying selected color as rounded view by tap.
/// Handle tap gesture to add UIView on canvas.
/// Handle event when undo & redo stack of UIView's was changed.
///
/// - Parameters
///     - *canvas*: UIView for adding selected color view by user's tap.
///     - *wasChanged*: Completion handler event when undo & redo stack of UIView's was changed.
class ColorCanvasHandler {
    private var canvas: UIView
    private var wasChangedCompletionHandler: (Bool, Bool) -> ()
    
    private var selectedColor: UIColor?
    
    private var hash: [UIColor? : Int] = [:]
    private var undoStack = Stack<UIView>()
    private var redoStack = Stack<UIView>()
    
    private var isUndoEnabled: Bool {
        self.undoStack.isEmpty ? false : true
    }
    
    private var isRedoEnabled: Bool {
        self.redoStack.isEmpty ? false : true
    }
    
    init(canvas: UIView,
         wasChanged: @escaping (Bool, Bool) -> ()) {
        self.canvas = canvas
        self.wasChangedCompletionHandler = wasChanged
        
        self.addTapGestureRecognizerToCanvas()
    }
    
    private func addTapGestureRecognizerToCanvas() {
        let tap = UITapGestureRecognizer(target: self, action: #selector(addView(gesture:)))
        self.canvas.addGestureRecognizer(tap)
    }
    
    @objc
    private func addView(gesture: UITapGestureRecognizer) {
        guard let color = selectedColor else {
            return
        }
        // Get location on canvas
        let location = gesture.location(in: self.canvas)
        // Add view at location on canvas with specific color
        addView(at: location, color: color)
    }
    
    private func addView(at location: CGPoint, color: UIColor?) {
        let size = 50.0
        let view = UILabel(frame: CGRect.init(x: location.x - size * 0.5, y: location.y - size * 0.5, width: size, height: size))
        view.backgroundColor = color
        // Add rounded corners
        view.layer.cornerRadius = size * 0.5
        view.clipsToBounds = true
        // Set font
        view.font = UIFont.boldSystemFont(ofSize: 20)
        // Set text attributes
        view.textColor = .white
        view.textAlignment = .center
        // Update hash for added color
        let current = self.updateHash(color:view.backgroundColor, ascending: true)
        view.text = String(current)
        // Add to canvas
        self.canvas.addSubview(view)
        // Add view to "undo" stack
        self.undoStack.push(view)
        // Completion handler
        self.wasChangedCompletionHandler(isUndoEnabled, isRedoEnabled)
    }
    
    @discardableResult
    private func updateHash(color: UIColor?, ascending: Bool) -> Int {
        var current = hash[color] ?? 0
        current = ascending ?
        current + 1 :
        max(0, current - 1)
        hash[color] = current
        return current
    }
    
}

extension ColorCanvasHandler {
        
    func colorWasChanged(_ color: UIColor?) {
        self.selectedColor = color
    }
    
    func undo() {
        guard let view = self.undoStack.pop() else {
            return
        }
        // Update hash for removed color
        self.updateHash(color: view.backgroundColor, ascending: false)
        // Remove view from canvas
        view.removeFromSuperview()
        // Add view to "redo" stack
        redoStack.push(view)
        // Completion handler
        self.wasChangedCompletionHandler(isUndoEnabled, isRedoEnabled)
    }
    
    func redo() {
        guard let view = self.redoStack.pop() else {
            return
        }
        // Get location from restored view on canvas
        let location = CGPoint(x: view.frame.origin.x + view.frame.width * 0.5,
                               y: view.frame.origin.y + view.frame.height * 0.5)
        // Add restored view at location on canvas with specific color
        addView(at: location, color: view.backgroundColor)
    }
    
}

